File: demos/05-cli-sdk/02-sdk/03-multi-agent/multi-agent-solution-node/coordinator.ts
import {
  CopilotClient,
  defineTool,
  type CopilotSession,
  type SessionConfig,
} from "@github/copilot-sdk";

const MODEL = "gpt-5-mini";

const listChanges = defineTool<{ since: string }>("list_changes", {
  description: "List the merged changes in the repository since a given tag",
  skipPermission: true,
  parameters: {
    type: "object",
    properties: {
      since: {
        type: "string",
        description: "The tag to list changes since",
      },
    },
    required: ["since"],
  },
  handler: async (args) => ({
    since: args.since,
    changes: [
      { id: "a91f2c", area: "cli", summary: "add --fleet to the prompt runner" },
      { id: "4d02be", area: "sdk", summary: "session.factory API for registered factories" },
      { id: "7c1188", area: "sdk", summary: "customAgents accepted at session creation" },
      { id: "be40aa", area: "docs", summary: "correct the tool permission defaults" },
    ],
  }),
});

async function runSpecialist(
  session: CopilotSession,
  label: string,
  prompt: string,
): Promise<string> {
  process.stdout.write(`\n--- ${label} ---\n`);
  const response = await session.sendAndWait({ prompt }, 180_000);
  const text = response?.data.content ?? "";
  console.log(text);
  return text;
}

async function main() {
  const client = new CopilotClient();

  const specialist = (systemMessage: string, tools: SessionConfig["tools"] = []) =>
    client.createSession({
      model: MODEL,
      streaming: false,
      tools,
      availableTools: ["custom:*"],
      systemMessage: { mode: "replace", content: systemMessage },
    });

  const researcher = await specialist(
    "You are a research specialist. Call the tools you are given, then report only the facts you retrieved as a compact bullet list. Never write prose around them.",
    [listChanges],
  );

  const builder = await specialist(
    "You are a release-notes writer. Turn the facts you are handed into a markdown section titled '## Unreleased', one bullet per change, grouped by area. Output the markdown only.",
  );

  const reviewer = await specialist(
    "You are a style reviewer. The house style forbids em dashes and requires every bullet to name its area. Answer with a verdict line 'VERDICT: PASS' or 'VERDICT: FAIL', then one line per violation.",
  );

  const facts = await runSpecialist(
    researcher,
    "researcher",
    "List the changes since tag v2.1 and report them.",
  );

  const draft = await runSpecialist(
    builder,
    "builder",
    `Write the release-notes section from these facts:\n\n${facts}`,
  );

  await runSpecialist(
    reviewer,
    "reviewer",
    `Review this release-notes section against the house style:\n\n${draft}`,
  );

  await client.stop();
  process.exit(0);
}

main();

File: demos/05-cli-sdk/02-sdk/03-multi-agent/multi-agent-solution-py/coordinator.py
import asyncio

from pydantic import BaseModel, Field

from copilot import CopilotClient, CopilotSession, Tool, define_tool

MODEL = "gpt-5-mini"


class ListChangesParams(BaseModel):
    since: str = Field(description="The tag to list changes since")


@define_tool(
    "list_changes",
    description="List the merged changes in the repository since a given tag",
    skip_permission=True,
)
async def list_changes(params: ListChangesParams) -> dict:
    return {
        "since": params.since,
        "changes": [
            {"id": "a91f2c", "area": "cli", "summary": "add --fleet to the prompt runner"},
            {"id": "4d02be", "area": "sdk", "summary": "session.factory API for registered factories"},
            {"id": "7c1188", "area": "sdk", "summary": "customAgents accepted at session creation"},
            {"id": "be40aa", "area": "docs", "summary": "correct the tool permission defaults"},
        ],
    }


async def run_specialist(session: CopilotSession, label: str, prompt: str) -> str:
    print(f"\n--- {label} ---")
    response = await session.send_and_wait(prompt, timeout=180.0)
    text = response.data.content if response else ""
    print(text)
    return text


async def main():
    client = CopilotClient()
    await client.start()

    async def specialist(system_message: str, tools: list[Tool] | None = None) -> CopilotSession:
        return await client.create_session(
            model=MODEL,
            streaming=False,
            tools=tools or [],
            available_tools=["custom:*"],
            system_message={"mode": "replace", "content": system_message},
        )

    researcher = await specialist(
        "You are a research specialist. Call the tools you are given, then report only the facts you retrieved as a compact bullet list. Never write prose around them.",
        [list_changes],
    )

    builder = await specialist(
        "You are a release-notes writer. Turn the facts you are handed into a markdown section titled '## Unreleased', one bullet per change, grouped by area. Output the markdown only.",
    )

    reviewer = await specialist(
        "You are a style reviewer. The house style forbids em dashes and requires every bullet to name its area. Answer with a verdict line 'VERDICT: PASS' or 'VERDICT: FAIL', then one line per violation.",
    )

    facts = await run_specialist(
        researcher,
        "researcher",
        "List the changes since tag v2.1 and report them.",
    )

    draft = await run_specialist(
        builder,
        "builder",
        f"Write the release-notes section from these facts:\n\n{facts}",
    )

    await run_specialist(
        reviewer,
        "reviewer",
        f"Review this release-notes section against the house style:\n\n{draft}",
    )

    await client.stop()


asyncio.run(main())
